packages/api/src/objectSet/ObjectSet.ts

```diff
@@ -254,4 +261,36 @@ export interface ObjectSet<
     listener: ObjectSetListener<Q, P>,
     opts?: ObjectSetListenerOptions<Q, P>,
   ) => { unsubscribe: () => void };
+
+  readonly withProperties: <
+    D extends WithPropertiesClause<Q>,
+  >(
+    clause: D,
+  ) => ObjectSet<
+    WithPropertiesObjectDefinition<
+      Q,
+      D
+    >
+  >;
 }
+
+type WithPropertiesObjectDefinition<
+  K extends ObjectOrInterfaceDefinition,
+  D extends WithPropertiesClause<K>,
+> = {
+  __DefinitionMetadata: {
+    properties: {
```

Comment:
This is all it seems like we need to propagate the types through all further object set operations. The FLUP here is to make the type of the whole object set extractable and definable so that a user can pass around object sets that are correctly types and do not have to be inferred.